nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java

```diff
@@ -773,6 +770,47 @@ private Object coerceStringToLong(final String fieldName, final String stringVal
                 : stringValue;
     }
 
+    private Record cloneRecord(final Record record) {
+        final Map<String, Object> valuesCopy = cloneValues(record.toMap());
+        return new MapRecord(record.getSchema(), valuesCopy, record.isTypeChecked(), record.isDropUnknownFields());
+    }
+
+    private Map<String, Object> cloneValues(final Map<String, Object> values) {
+        final Map<String, Object> cloned = new LinkedHashMap<>(values.size());
+        for (final Map.Entry<String, Object> entry : values.entrySet()) {
+            cloned.put(entry.getKey(), cloneValue(entry.getValue()));
+        }
+        return cloned;
+    }
+
+    private Object cloneValue(final Object value) {
```

Comment:
Instead of creating all these clone methods, it would be cleaner to use something like Apache Commons `SerializationUtils#clone`. The only thing you would have to make sure though is all of the data to be cloned implements the Serializable interface.

Reply:
In our case the records are `MapRecord` instances, and `MapRecord` (and other `Record` implementations) do not implement `Serializable`. Besides, Java serialization would add noticeable overhead on a hot path that can process large batches, whereas the current approach only walk the map/collection structures we actually store.